offscreen_effect_player/yuv_converter.cpp

```diff
@@ -0,0 +1,356 @@
+        output.data = std::make_unique<uint8_t[]>(output.size);
+    }
+
+    /* just in case, disable dropping geometry */
+    glDisable(GL_CULL_FACE);
+
+    /* bind drawing geometry */
+    glBindVertexArray(m_vao);
+
+    /* use shader and send texture matrix to shader program */
+    m_shader.use();
+
+    /* bind input texture */
+    glActiveTexture(GL_TEXTURE0);
+    glBindTexture(GL_TEXTURE_2D, gl_texture);
+    glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MIN_FILTER, GL_LINEAR);
+    glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MAG_FILTER, GL_LINEAR);
+    m_shader.set_uniform("in_texture", 0);
+
+    /* pixel step used in the shader to access neighboring pixels */
+    m_shader.set_uniform("pixel_step", m_pixel_step_y[0], m_pixel_step_y[1]);
+    /* render Y plane to the framebuffer*/
+    glBindFramebuffer(GL_FRAMEBUFFER, m_fbo.fbo);
+    m_shader.set_uniform("plane_coef", m_y_plane_coefs[0], m_y_plane_coefs[1],
+            m_y_plane_coefs[2], m_y_plane_coefs[3]);
+    glViewport(0, 0, m_fbo.width, m_height);
+    glDrawArrays(GL_TRIANGLE_STRIP, m_draw_indent, drawing_plane_vert_count);
+
+    /* pixel step used in the shader to access neighboring pixels */
+    m_shader.set_uniform("pixel_step", m_pixel_step_uv[0], m_pixel_step_uv[1]);
+
+    /* render U and V planes to the framebuffer */
+    m_shader.set_uniform("plane_coef", m_u_plane_coefs[0], m_u_plane_coefs[1],
+            m_u_plane_coefs[2], m_u_plane_coefs[3]);
+    glViewport(0, m_height, half_viewport_width, half_height);
+    glDrawArrays(GL_TRIANGLE_STRIP, m_draw_indent, drawing_plane_vert_count);
+    m_shader.set_uniform("plane_coef", m_v_plane_coefs[0], m_v_plane_coefs[1],
+            m_v_plane_coefs[2], m_v_plane_coefs[3]);
+    glViewport(half_viewport_width, m_height, half_viewport_width, half_height);
+    glDrawArrays(GL_TRIANGLE_STRIP, m_draw_indent, drawing_plane_vert_count);
+    /* and read all YUV planes data */
+    glReadPixels(0, 0, m_fbo.width, m_fbo.height, GL_RGBA, GL_UNSIGNED_BYTE, output.data.get());
+
+    /* unbind all */
+    glBindFramebuffer(GL_FRAMEBUFFER, 0);
+    glBindVertexArray(0);
+    glBindTexture(GL_TEXTURE_2D, 0);
+    m_shader.unuse();
+
+    output.y_plane_data = output.data.get();
+    output.u_plane_data = output.data.get() + stride * m_height;
+    output.v_plane_data = output.data.get() + stride * m_height + half_viewport_width * 4;
+    output.y_plane_stride = stride;
+    output.u_plane_stride = stride;
+    output.v_plane_stride = stride;
+}
+
+/* yuv_converter::update_pixel_steps */
+void yuv_converter::update_pixel_steps()
+{
+    if (m_width == 0) {
+        return;
+    }
+    float x_step {0.0f};
+    float y_step {0.0f};
+    switch (m_rotation) {
+        case rotation::deg_0:
+            x_step = 1.0f / m_width;
+            break;
+        case rotation::deg_90:
+            y_step = (m_vertical_flip ? 1.0f : -1.0f) / m_width;
+            break;
+        case rotation::deg_180:
+            x_step = -1.0f / m_width;

```

Comment:
Check the spikes on the resulted image with that value of step for this orientation - I see spikes in my case and fixed that by removing - sign.